File: javascript/packages/linter/src/rules/html-tag-name-lowercase.ts
import { ParserRule, BaseAutofixContext, Mutable } from "../types.js"
import { BaseRuleVisitor, findParent } from "./rule-utils.js"
import { isNode, getTagName, HTMLOpenTagNode, isHTMLElementNode } from "@herb-tools/core"

import type { UnboundLintOffense, LintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { HTMLElementNode, HTMLCloseTagNode, ParseResult, XMLDeclarationNode, Node } from "@herb-tools/core"

interface TagNameAutofixContext extends BaseAutofixContext {
  node: Mutable<HTMLOpenTagNode | HTMLCloseTagNode>
  tagName: string
  correctedTagName: string
}

class XMLDeclarationChecker extends BaseRuleVisitor {
  hasXMLDeclaration: boolean = false

  visitXMLDeclarationNode(_node: XMLDeclarationNode): void {
    this.hasXMLDeclaration = true
  }

  visitChildNodes(node: Node): void {
    if (this.hasXMLDeclaration) return
    super.visitChildNodes(node)
  }
}

class TagNameLowercaseVisitor extends BaseRuleVisitor<TagNameAutofixContext> {
  visitHTMLElementNode(node: HTMLElementNode): void {
    if (getTagName(node).toLowerCase() === "svg") {
      this.checkTagName(node.open_tag)
      this.checkTagName(node.close_tag)
    } else {
      super.visitHTMLElementNode(node)
    }
  }

  visitHTMLOpenTagNode(node: HTMLOpenTagNode) {
    this.checkTagName(node)
  }

  visitHTMLCloseTagNode(node: HTMLCloseTagNode) {
    this.checkTagName(node)
  }

  private checkTagName(node: HTMLOpenTagNode | HTMLCloseTagNode | null): void {
    if (!node) return

    const tagName = getTagName(node)

    if (!tagName) return

    const lowercaseTagName = tagName.toLowerCase()

    const type = isNode(node, HTMLOpenTagNode) ? "Opening" : "Closing"
    const open = isNode(node, HTMLOpenTagNode) ? "<" : "</"

    if (tagName !== lowercaseTagName) {
      this.addOffense(
        `${type} tag name \`${open}${tagName}>\` should be lowercase. Use \`${open}${lowercaseTagName}>\` instead.`,
        node.tag_name!.location,
        {
          node,
          tagName,
          correctedTagName: lowercaseTagName
        }
      )
    }
  }
}

export class HTMLTagNameLowercaseRule extends ParserRule<TagNameAutofixContext> {
  static autocorrectable = true
  name = "html-tag-name-lowercase"

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "error",
      exclude: ["**/*.xml", "**/*.xml.erb"]
    }
  }

  isEnabled(result: ParseResult, _context?: Partial<LintContext>): boolean {
    const checker = new XMLDeclarationChecker(this.name)

    checker.visit(result.value)

    return !checker.hasXMLDeclaration
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense<TagNameAutofixContext>[] {
    const visitor = new TagNameLowercaseVisitor(this.name, context)

    visitor.visit(result.value)

    return visitor.offenses
  }

  autofix(offense: LintOffense<TagNameAutofixContext>, result: ParseResult, _context?: Partial<LintContext>): ParseResult | null {
    if (!offense.autofixContext) return null

    const { node, correctedTagName } = offense.autofixContext
    if (!node.tag_name) return null

    node.tag_name.value = correctedTagName

    const parentElement = findParent(result.value, node as any as Node)
    if (!parentElement || !isHTMLElementNode(parentElement)) return result

    switch (node.type) {
      case "AST_HTML_OPEN_TAG_NODE":
        if (!parentElement.close_tag) break

        const closeTag = parentElement.close_tag as Mutable<HTMLCloseTagNode>
        closeTag.tag_name!.value = correctedTagName
        break
      case "AST_HTML_CLOSE_TAG_NODE":
        const openTag = parentElement.open_tag as Mutable<HTMLOpenTagNode>
        openTag.tag_name!.value = correctedTagName
        break
      default:
        break
    }

    return result
  }
}

File: rust/herb-linter/src/rules/html_tag_name_lowercase.rs
use crate::autofix::{close_tag_name, for_each_element_mut, location_matches};
use crate::offense::Offense;
use crate::offense::UnboundOffense;
use crate::rule::{LintContext, ParserRule, Rule};
use crate::utils::tag_utils::get_open_tag;
use crate::utils::tag_utils::{get_tag_name_from_open_tag, tag_name_location};
use herb::union_types::*;

use herb::nodes::*;
use herb::ParseResult;
use herb::Visitor;
use herb_config::{Severity, SeverityConfig};

struct XMLDeclarationChecker {
  has_xml_declaration: bool,
}

impl Visitor for XMLDeclarationChecker {
  fn visit_xml_declaration_node(&mut self, _node: &XMLDeclarationNode) {
    self.has_xml_declaration = true;
  }
}

rule_visitor!(TagNameLowercaseVisitor);

impl TagNameLowercaseVisitor {
  fn check_open_tag(&mut self, node: &HTMLOpenTagNode) {
    if let Some(tag_name) = get_tag_name_from_open_tag(node) {
      let lowercase = tag_name.to_lowercase();
      if tag_name != lowercase {
        let location = tag_name_location(node);

        self.add_offense(
          format!("Opening tag name `<{}>` should be lowercase. Use `<{}>` instead.", tag_name, lowercase),
          location,
        );
      }
    }
  }

  fn check_close_tag(&mut self, node: &HTMLCloseTagNode) {
    if let Some(token) = &node.tag_name {
      let tag_name = &token.value;
      let lowercase = tag_name.to_lowercase();

      if tag_name.as_str() != lowercase {
        self.add_offense(
          format!("Closing tag name `</{}>` should be lowercase. Use `</{}>` instead.", tag_name, lowercase),
          token.location.clone(),
        );
      }
    }
  }
}

impl Visitor for TagNameLowercaseVisitor {
  fn visit_html_element_node(&mut self, node: &HTMLElementNode) {
    // For SVG elements, only check the SVG tag itself (not children)
    if let Some(tag_name) = node.tag_name.as_ref().map(|token| token.value.as_str()) {
      if tag_name.eq_ignore_ascii_case("svg") {
        if let Some(open_tag) = crate::utils::tag_utils::get_open_tag(node) {
          self.check_open_tag(open_tag);
        }
        if let Some(herb::union_types::ERBEndNodeOrHTMLCloseTagNodeOrHTMLOmittedCloseTagNodeOrHTMLVirtualCloseTagNode::HTMLCloseTagNode(close)) =
          &node.close_tag
        {
          self.check_close_tag(close);
        }
        return;
      }
    }
    self.walk_html_element_node(node);
  }

  fn visit_html_open_tag_node(&mut self, node: &HTMLOpenTagNode) {
    self.check_open_tag(node);
  }

  fn visit_html_close_tag_node(&mut self, node: &HTMLCloseTagNode) {
    self.check_close_tag(node);
  }
}

pub struct HTMLTagNameLowercaseRule;

impl Rule for HTMLTagNameLowercaseRule {
  fn has_autofix(&self) -> bool {
    true
  }

  fn autocorrectable(&self) -> bool {
    true
  }

  fn name(&self) -> &'static str {
    "html-tag-name-lowercase"
  }

  fn introduced_in(&self) -> Option<&'static str> {
    Some("0.4.0")
  }

  fn default_severity(&self) -> SeverityConfig {
    SeverityConfig::Severity(Severity::Error)
  }

  fn default_exclude(&self) -> &[&str] {
    &["**/*.xml", "**/*.xml.erb"]
  }
}

impl ParserRule for HTMLTagNameLowercaseRule {
  fn check(&self, result: &ParseResult, _context: &LintContext) -> Vec<UnboundOffense> {
    // Skip if the file contains an XML declaration
    let mut xml_checker = XMLDeclarationChecker { has_xml_declaration: false };
    xml_checker.visit_document_node(&result.value);
    if xml_checker.has_xml_declaration {
      return Vec::new();
    }

    let mut visitor = TagNameLowercaseVisitor::new(self.name());
    visitor.visit_document_node(&result.value);
    visitor.offenses
  }
  fn autofix(&self, offense: &Offense, document: &mut DocumentNode, _context: &LintContext) -> bool {
    let mut fixed = false;

    for_each_element_mut(document, &mut |element| {
      let open_matches = get_open_tag(element)
        .and_then(|open_tag| open_tag.tag_name.as_ref())
        .map(|token| location_matches(&token.location, offense))
        .unwrap_or(false);

      let close_matches = close_tag_name(element).map(|token| location_matches(&token.location, offense)).unwrap_or(false);

      if !open_matches && !close_matches {
        return;
      }

      let corrected = if open_matches {
        get_open_tag(element)
          .and_then(|tag| tag.tag_name.as_ref())
          .map(|token| token.value.to_lowercase())
      } else {
        close_tag_name(element).map(|token| token.value.to_lowercase())
      };

      let corrected = match corrected {
        Some(corrected) => corrected,
        None => return,
      };

      if let Some(ERBOpenTagNodeOrHTMLConditionalOpenTagNodeOrHTMLOpenTagNode::HTMLOpenTagNode(open_tag)) = element.open_tag.as_mut() {
        if let Some(token) = open_tag.tag_name.as_mut() {
          token.value = corrected.clone();
        }
      }

      if let Some(ERBEndNodeOrHTMLCloseTagNodeOrHTMLOmittedCloseTagNodeOrHTMLVirtualCloseTagNode::HTMLCloseTagNode(close_tag)) = element.close_tag.as_mut() {
        if let Some(token) = close_tag.tag_name.as_mut() {
          token.value = corrected;
        }
      }

      fixed = true;
    });

    fixed
  }
}
